### DISCUSSION TITLE
How to implement global error handling for useAsyncData?

### DISCUSSION THREAD

**Discussion by TimGuendel** [nuxt/nuxt#28357] [Questions]
So I am using a custom [ofetch][1] instance that is defined inside a Nuxt Plugin. It has its own error handling (simplified for this example):

    import type {FetchOptions} from "ofetch";
    
    export default defineNuxtPlugin(() => {
        const api = $fetch.create({
            async onResponseError({ response }) {
                // [...]
                showError({
                    statusCode,
                    fatal: true,
                });
            },
        });
    
        const get = async <T>(url: string, options?: FetchOptions) => {
            return await api<T>(url, { ...options, method: 'GET' });
        };
    
        return {
            provide: {
                apiGet: get,
            },
        };
    });

I can now use it in my app. However, when using it inside [useAsyncData][2], the error handling **does not work if the error occurs when rendered on the server side**.

This is because [showError][3] only calls the app:error hook only if executed [on client side ][4].

If I replace showError with [createError][5], the created error is returned by useAsyncData and I can handle it there - but this is defeating my global error handling approach.

**I don't want to be handling the error in every component**. How can I implement this without duplicating my code? So far, this is the workaround I am working with:

    export function handleErrors(...errors: (NuxtError | undefined | null)[]) {
        errors.forEach((error) => {
            if (error) {
                throw createError(error);
            }
        });
    }

And its usage:

    const { error } = await useAsyncData(() => $apiGet('/wrong/url'));
    handleErrors(error.value);

which feels wonky.

  [1]: https://github.com/unjs/ofetch
  [2]: https://nuxt.com/docs/api/composables/use-async-data
  [3]: https://nuxt.com/docs/api/utils/show-error
  [4]: https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/app/composables/error.ts#L27
  [5]: https://nuxt.com/docs/api/utils/create-error

**Comment by sirgeon** [comment]
To handle errors globally in your Nuxt app without duplicating code in every component, update your custom plugin to handle both server-side and client-side errors, and use middleware for consistent error management.

```ts
import type { FetchOptions } from 'ofetch';
import { createError, showError } from 'nuxt/app';

export default defineNuxtPlugin(() => {
    const api = $fetch.create({
        async onResponseError({ response }) {
            const statusCode = response.status;

            if (process.server) {
                throw createError({
                    statusCode,
                    message: response.statusText,
                });
            } else {
                showError({
                    statusCode,
                    fatal: true,
                });
            }
        },
    });

    const get = async <T>(url: string, options?: FetchOptions) => {
        return await api<T>(url, { ...options, method: 'GET' });
    };

    return {
        provide: {
            apiGet: get,
        },
    };
});
```
global middleware for catching errors:
```ts
export default defineNuxtRouteMiddleware((to, from) => {
    const { error } = useNuxtApp();

    if (error.value) {
        // Handle the error globally here, e.g., redirect to a custom error page
        return navigateTo('/error');
    }
});
```

u can then register this globally in ur nuxt conf

```ts
export default defineNuxtConfig({
    router: {
        middleware: 'errorHandler'
    }
});
```

hope this helps u :)

**Reply by TimGuendel** [reply]
Do you mean `const error = useError();` instead of `const { error } = useNuxtApp();` ?
Also, no need to register the middleware in the config if you name the file .global.ts ✅
I will try this, thanks. But will the middelware be triggered when no route change occurs? 🤔

**Reply by rolfkoizumi** [reply]
It's weird. `useError()` somehow never contains an error for me when used in a middleware in combination with `useAsyncData()` and a modified `onResponseError()`.

(`const { error } = useNuxtApp();` does not exist for me)

I have created a minimal reproduction here: https://stackblitz.com/edit/github-bjbded?file=middleware%2FerrorHandling.global.ts
The second link "useAsyncData error" will open a page that shows the error-page client-side. But when you hit refresh, the error-page won't be triggered.

But even if this approach worked. The middleware is triggered on server & client side. On server-side, it is triggered **before** the fetch&render, so it won't be able to catch an error.
On client side then, it will be triggered after server-side-render and hydration. If an error is found then, the error page could be shown.
The bad thing is that this means the server won't return the error-page at this point, which means:
- The page might flicker, showing the rendered page (with potentially missing parts) before showing the error page
  - We might encounter render-errors because data is missing
- Google/Search engines might not get the error-page if they don't execute javascript when indexing

**Reply by zc16607** [reply]
I also encountered the same problem. 
So, this method is not feasible. Is there any other solution?
